File: Dot examples/ASP Book Examples/Book Examples/Chapter29/VB/App_Code/Pager.vb
Imports System
Imports System.Web.UI
Imports System.Web.UI.WebControls

Namespace myControls
    Public Class Pager
        Inherits WebControl
        Implements IPostBackEventHandler

        Private _controlToPage As String

        Public Property ControlToPage() As String
            Get
                Return _controlToPage
            End Get
            Set(ByVal Value As String)
                _controlToPage = value
            End Set
        End Property

        Protected Overrides Sub RenderContents(ByVal writer As HtmlTextWriter)
            Dim grid As GridView = GetControlToPage()

            Dim i As Integer
            For i = 0 To grid.PageCount - 1 Step i + 1
                Dim eRef As String = Page.ClientScript.GetPostBackClientHyperlink(Me, i.ToString())
                writer.Write("[")
                If i = grid.PageIndex Then
                    writer.AddStyleAttribute(HtmlTextWriterStyle.FontWeight, "bold")
                End If
                writer.AddAttribute(HtmlTextWriterAttribute.Href, eRef)
                writer.RenderBeginTag(HtmlTextWriterTag.A)
                writer.Write("{0}", i + 1)
                writer.RenderEndTag()
                writer.Write("] ")
            Next
        End Sub

        Private Function GetControlToPage() As GridView
            If String.IsNullOrEmpty(_controlToPage) Then
                Throw New Exception("Must set ControlToPage property")
            End If
            Return CType(Page.FindControl(_controlToPage), GridView)
        End Function

        Public Sub RaisePostBackEvent(ByVal eventArgument As String) Implements IPostBackEventHandler.RaisePostBackEvent
            Dim grid As GridView = GetControlToPage()
            grid.PageIndex = Int32.Parse(eventArgument)
        End Sub
    End Class
End Namespace

File: Dot examples/ASP Book Examples/Book Examples/Chapter29/CS/App_Code/Pager.cs
using System;
using System.Web.UI;
using System.Web.UI.WebControls;

namespace myControls
{
    public class Pager : WebControl, IPostBackEventHandler
    {
        string _controlToPage;

        public string ControlToPage
        {
            get { return _controlToPage; }
            set { _controlToPage = value; }
        }

        protected override void RenderContents(HtmlTextWriter writer)
        {
            GridView grid = GetControlToPage();

            for (int i = 0; i < grid.PageCount; i++)
            {
                string eRef = Page.ClientScript.GetPostBackClientHyperlink(this, i.ToString());
                writer.Write("[");
                if (i == grid.PageIndex)
                    writer.AddStyleAttribute(HtmlTextWriterStyle.FontWeight, "bold");
                writer.AddAttribute(HtmlTextWriterAttribute.Href, eRef);
                writer.RenderBeginTag(HtmlTextWriterTag.A);
                writer.Write("{0}", i + 1);
                writer.RenderEndTag();
                writer.Write("] ");
            }
        }

        private GridView GetControlToPage()
        {
            if (String.IsNullOrEmpty(_controlToPage))
                throw new Exception("Must set ControlToPage property");
            return (GridView)Page.FindControl(_controlToPage);
        }

        public void RaisePostBackEvent(string eventArgument)
        {
            GridView grid = GetControlToPage();
            grid.PageIndex = Int32.Parse(eventArgument);
        }
    }
}